src/test/versionbits_tests.cpp

```diff
@@ -0,0 +1,185 @@
+                           .Mine(3000, TestTime(30005), 0x100).TestFailed()
+
+        // DEFINED -> STARTED -> FAILED
+                           .Reset().TestDefined()
+                           .Mine(1, TestTime(1), 0).TestDefined()
+                           .Mine(1000, TestTime(10000) - 1, 0x100).TestDefined() // One second more and it would be defined
+                           .Mine(2000, TestTime(10000), 0x100).TestStarted() // So that's what happens the next period
+                           .Mine(2051, TestTime(10010), 0).TestStarted() // 51 old blocks
+                           .Mine(2950, TestTime(10020), 0x100).TestStarted() // 899 new blocks
+                           .Mine(3000, TestTime(20000), 0).TestFailed() // 50 old blocks (so 899 out of the past 1000)
+                           .Mine(4000, TestTime(20010), 0x100).TestFailed()
+
+        // DEFINED -> STARTED -> FAILED while threshold reached
+                           .Reset().TestDefined()
+                           .Mine(1, TestTime(1), 0).TestDefined()
+                           .Mine(1000, TestTime(10000) - 1, 0x101).TestDefined() // One second more and it would be defined
+                           .Mine(2000, TestTime(10000), 0x101).TestStarted() // So that's what happens the next period
+                           .Mine(2999, TestTime(30000), 0x100).TestStarted() // 999 new blocks
+                           .Mine(3000, TestTime(30000), 0x100).TestFailed() // 1 new block (so 1000 out of the past 1000 are new)
+                           .Mine(3999, TestTime(30001), 0).TestFailed()
+                           .Mine(4000, TestTime(30002), 0).TestFailed()
+                           .Mine(14333, TestTime(30003), 0).TestFailed()
+                           .Mine(24000, TestTime(40000), 0).TestFailed()
+
+        // DEFINED -> STARTED -> LOCKEDIN at the last minute -> ACTIVE
+                           .Reset().TestDefined()
+                           .Mine(1, TestTime(1), 0).TestDefined()
+                           .Mine(1000, TestTime(10000) - 1, 0x101).TestDefined() // One second more and it would be defined
+                           .Mine(2000, TestTime(10000), 0x101).TestStarted() // So that's what happens the next period
+                           .Mine(2050, TestTime(10010), 0x200).TestStarted() // 50 old blocks
+                           .Mine(2950, TestTime(10020), 0x100).TestStarted() // 900 new blocks
+                           .Mine(2999, TestTime(19999), 0x200).TestStarted() // 49 old blocks
+                           .Mine(3000, TestTime(29999), 0x200).TestLockedIn() // 1 old block (so 900 out of the past 1000)
+                           .Mine(3999, TestTime(30001), 0).TestLockedIn()
+                           .Mine(4000, TestTime(30002), 0).TestActive()
+                           .Mine(14333, TestTime(30003), 0).TestActive()
+                           .Mine(24000, TestTime(40000), 0).TestActive();
+    }
+}
```

Comment:
I wrote up a couple additional tests to try sanity checking the mainnet deployments:

```
diff --git a/src/test/versionbits_tests.cpp b/src/test/versionbits_tests.cpp
index 9de8461..c384f92 100644
--- a/src/test/versionbits_tests.cpp
+++ b/src/test/versionbits_tests.cpp
@@ -6,6 +6,7 @@
 #include "random.h"
 #include "versionbits.h"
 #include "test/test_bitcoin.h"
+#include "chainparams.h"
 
 #include <boost/test/unit_test.hpp>
 
@@ -180,6 +181,28 @@ BOOST_AUTO_TEST_CASE(versionbits_test)
                            .Mine(14333, TestTime(30003), 0).TestActive()
                            .Mine(24000, TestTime(40000), 0).TestActive();
     }
+
+    // Sanity checks of version bit deployments
+    const Consensus::Params &mainnetParams = Params(CBaseChainParams::MAIN).GetConsensus();
+    for (int i=0; i<(int) Consensus::MAX_VERSION_BITS_DEPLOYMENTS; i++) {
+        uint32_t bitmask = VersionBitsMask(mainnetParams, (Consensus::DeploymentPos)i);
+        // Make sure that no deployment tries to set an invalid bit.
+        BOOST_CHECK(bitmask < (uint32_t)VERSIONBITS_TOP_BITS);
+
+        // Verify that the deployment windows of different deployment using the
+        // same bit are disjoint.
+        // This test may need modification at such time as a new deployment
+        // is proposed that reuses the bit of an activated soft fork, before the
+        // end time of that soft fork.  (Alternatively, the end time of that
+        // activated soft fork could be later changed to be earlier to avoid
+        // overlap.)
+        for (int j=i+1; j<(int) Consensus::MAX_VERSION_BITS_DEPLOYMENTS; j++) {
+            if (VersionBitsMask(mainnetParams, (Consensus::DeploymentPos)j) == bitmask) {
+                BOOST_CHECK(mainnetParams.vDeployments[j].nStartTime > mainnetParams.vDeployments[i].nTimeout ||
+                        mainnetParams.vDeployments[i].nStartTime > mainnetParams.vDeployments[j].nTimeout);
+            }
+        }
+    }
 }
 
 BOOST_AUTO_TEST_SUITE_END()
```

Perhaps worth including something like that here?